### PR TITLE
2513: Rename api-client to shared and open a new subfolder api

### PR DESCRIPTION
**Pull request by lunars97**
### Short description

Api-client has grown a lot and it does not only contain things related to api-client, so it was decided to rename it to `shared` and create a subfolder with the name `api` where it will contain api related files and folders

### Proposed changes
- Rename api-client to shared
- Create a subfolder called api

Fixes: #2513 

---

### CODE REVIEW

**Comment by sarahsporck** [comment]
What do you think of having two entry points? `import ... from 'shared'` and `import ... from 'shared/api-client'`? This could make the imports a lot cleaner I think.

**Comment by steffenkleinle** [comment]
> What do you think of having two entry points? `import ... from 'shared'` and `import ... from 'shared/api-client'`? This could make the imports a lot cleaner I think.

If thats easy to implement it sounds nice. I wouldn't mind keeping the current as well though. I would use `shared/api` though with ommitting of the `client`.

**Comment by lunars97** [comment]
> What do you think of having two entry points? `import ... from 'shared'` and `import ... from 'shared/api-client'`? This could make the imports a lot cleaner I think.

By entry points do you mean what is in main in shared/package.json?

**Comment by lunars97** [comment]
> I think at least errors, endpoints, mapping and models subfolders should also go to `api` subfolder as well as some of the folders that are now just in the root folder of `shared`, e.g. `mapAvailableLanguages`, MapParamsToBody`, ... as those are all part of fetching stuff from our api.

Do you mean all the subfolders and files in src folder?

**Comment by steffenkleinle** [comment]
> > I think at least errors, endpoints, mapping and models subfolders should also go to `api` subfolder as well as some of the folders that are now just in the root folder of `shared`, e.g. `mapAvailableLanguages`, MapParamsToBody`, ... as those are all part of fetching stuff from our api.
> 
> Do you mean all the subfolders and files in src folder?

No, not all of them, just the ones that are for handling the api. The examples above are all doing that.

**Comment by sarahsporck** [comment]
So to visualize sth like 
```
- shared
  - @types/
  - api
    - endpoints/
    - errors/
    - mappings/
    - models/
    - MapParamsToBody.ts
    - MapParamsToUrlType.ts
    - MapResponseType
    - types.ts
    - index.ts // contains all exposed api exports
  - constants/
  - maps/
  - routes/
  - testing/
  - tracking/
  - utils/
  - mapAvailableLanguages.ts
  - normalizePath.ts
  - types.ts  // only contains type TimeSlots for now?
  - index.ts  // contains all shared exposed exports
```

I think the main field in the package.json can only list one file. This could be adjusted once the `shared` folder is an esm (#2404) package. If sth. does not work feel free to say sth.. Up till now this is also only a concept ^^

**Comment by steffenkleinle** [comment]
> So to visualize sth like
> 
> ```
> - shared
>   - @types/
>   - api
>     - endpoints/
>     - errors/
>     - mappings/
>     - models/
>     - MapParamsToBody.ts
>     - MapParamsToUrlType.ts
>     - MapResponseType
>     - types.ts
>     - index.ts // contains all exposed api exports
>   - constants/
>   - maps/
>   - routes/
>   - testing/
>   - tracking/
>   - utils/
>   - mapAvailableLanguages.ts
>   - normalizePath.ts
>   - types.ts  // only contains type TimeSlots for now?
>   - index.ts  // contains all shared exposed exports
> ```
> 
> I think the main field in the package.json can only list one file. This could be adjusted once the `shared` folder is an esm (#2404) package. If sth. does not work feel free to say sth.. Up till now this is also only a concept ^^



> So to visualize sth like
> 
> ```
> - shared
>   - @types/
>   - api
>     - endpoints/
>     - errors/
>     - mappings/
>     - models/
>     - MapParamsToBody.ts
>     - MapParamsToUrlType.ts
>     - MapResponseType
>     - types.ts
>     - index.ts // contains all exposed api exports
>   - constants/
>   - maps/
>   - routes/
>   - testing/
>   - tracking/
>   - utils/
>   - mapAvailableLanguages.ts
>   - normalizePath.ts
>   - types.ts  // only contains type TimeSlots for now?
>   - index.ts  // contains all shared exposed exports
> ```
> 
> I think the main field in the package.json can only list one file. This could be adjusted once the `shared` folder is an esm (#2404) package. If sth. does not work feel free to say sth.. Up till now this is also only a concept ^^

I mostly agree with your list, only I think `mapAvailableLanguages` should go to `api/mappings` and `normalizePath` should go to `utils`. Also, `MapParamsToBody/UrlType` and `MapResponseType` could go to `api/endpoints`.

**Review by steffenkleinle — COMMENTED** [review]
Looking good already. I have a few more requests if we are already adjusting the whole structure:
- Move `shared/api/mapAvailableLanguages` to `shared/api/mapping`
- Move `shared/api/maps/index` to `shared/constants` and rename to `map.ts`
- Move `shared/normalizePath` to `shared/utils`

Thank you! :pray: 

Otherwise a few imports have to be adjusted still, e.g. ts:check returns a lot of errors.

File: README.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ Both projects are whitelabelled with different [build configs](build-configs/REA
 There are also separate sub-projects used by both projects:
 
 - **[translations](translations/README.md)**: Translated texts and utilities to manage them
-- **[api-client](api-client/README.md)**: Implementation of the endpoints (v3) of the integreat cms
+- **[shared](shared/README.md)**: Implementation of the endpoints (v3) of the integreat cms
```
Comment (steffenkleinle) [inline README.md:42]: Maybe you can adjust the description a little/add that it also includes shared things (utils, constants, ...)...

**Review by steffenkleinle — COMMENTED** [review]
Hmmm, there are quite a few changes that are not related to the issue of renaming, e.g. in e2e-tests. Is there any reason on changing these files? Could we do that in a separate PR or revert that?

**Comment by lunars97** [comment]
I am getting test errors which I can not resolve, it is saying that 
`Test suite failed to run

    TypeError: api_1.PoiCategoryModel is not a constructor

      20 |     phoneNumber: '012345',
      21 |     email: 'test@example.com',
    > 22 |     category: new PoiCategoryModel({
         |               ^
      23 |       color: '#1DC6C6',
      24 |       iconName: 'gastronomy',
      25 |       id: 10,
`

File: shared/routes/RouteInformationTypes.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ import {
   SearchRouteType,
   SprungbrettOfferRouteType,
   TuNewsType,
-} from '.'
+} from './index'
```
Comment (LeandraH) [inline shared/routes/RouteInformationTypes.ts:17]: You don't need to add the `index` everywhere. An import from a directory automatically points to the `index` file, so `'.'` leads to exactly the same file as `'./index'`

File: shared/routes/index.ts [hunk 3]
```diff
@@ -69,9 +69,6 @@ export const IMAGE_VIEW_MODAL_ROUTE: ImageViewModalRouteType = 'image'
 export type FeedbackModalRouteType = 'feedback'
 export const FEEDBACK_MODAL_ROUTE: FeedbackModalRouteType = 'feedback'
 
-export type ConsentRouteType = 'consent'
-export const CONSENT_ROUTE: ConsentRouteType = 'consent'
-
```
Comment (LeandraH) [inline shared/routes/index.ts:74]: Why did you delete these?

Comment (lunars97) [reply shared/routes/index.ts:74]: merge error :(

File: .circleci/src/commands/unit_test.yml [hunk 4]
```diff
@@ -1,22 +1,8 @@
 description: Runs the test task and stores coverage and junit data
```
Comment (ztefanie) [inline .circleci/src/commands/unit_test.yml:1]: Are these changes on codeclimate on purpose? I think you accidentally deleted some of @sarahsporck changes?

File: release-notes/2023.12.1/2505-filter-button-too-close.yml
Comment (ztefanie) [inline release-notes/2023.12.1/2505-filter-button-too-close.yml:1]: Why have you deleted these release notes?

**Review by ztefanie — COMMENTED** [review]
If you want to revert the changes in the `.circleci` and `release-notes` folder you can just do:

```
git checkout origin/main -- release-notes/
git checkout origin/main -- .circleci/
```

and the git commit and git push.

File: release-notes/2023.12.1/2505-filter-button-too-close.yml
Comment (lunars97) [reply release-notes/2023.12.1/2505-filter-button-too-close.yml:1]: @ztefanie because it was not related to my PR?

Comment (lunars97) [reply release-notes/2023.12.1/2505-filter-button-too-close.yml:1]: @ztefanie ahh it should have been merged

Comment (steffenkleinle) [reply release-notes/2023.12.1/2505-filter-button-too-close.yml:1]: For that reason they should not be deleted. They are related to some other issue and not your PR, so they should stay the were before.

**Comment by steffenkleinle** [comment]
@ztefanie I think we are ready :) Can't rule out that we missed something but lets hope the best :crossed_fingers:

**Review by ztefanie — APPROVED** [review]
